teos/src/gatekeeper.rs

```diff
@@ -324,12 +306,15 @@ impl chain::Listen for Gatekeeper {
         log::info!("New block received: {}", header.block_hash());
 
         // Expired user deletion is delayed. Users are deleted when their subscription is outdated, not expired.
-        let outdated_users = self.get_outdated_user_ids(height);
+        let outdated_users = self.get_outdated_users(height);
         if !outdated_users.is_empty() {
-            self.registered_users
-                .lock()
-                .unwrap()
-                .retain(|id, _| !outdated_users.contains(id));
+            // Remove the outdated users from memory first.
+            {
+                let mut registered_users = self.registered_users.lock().unwrap();
+                for outdated_user in outdated_users.iter() {
+                    registered_users.remove(outdated_user);
+                }
```

Comment:
Ummm, `retain` goes over the whole hashmap and executes the given function against each (k, v) in it (+ `contains` isn't efficient on a vector).
`outdated_users` is generally gonna be way smaller than all the registered users.

Reply:
Revisiting this, I think we better remove the extra scope on `registered_users` to make things cleaner. Don't think it would degrade performance (and if so, for a little block connected time, and only if users are actually being deleted) since all API methods end up actually accessing the DB as well.